### PR TITLE
Support fallback syntax in .npmrc file

### LINKED ISSUE #5029: [rush] Rush should support the ${NAME:-fallback} syntax in .npmrc
## Summary

I want to set the .npmrc based on environment variables, so I added the following content to the .npmrc:
```ini
store-dir=${PNPM_STORE_DIR:-~/.pnpm/store}
```
This configuration means that when the environment variable PNPM_SOTRE_DIR exists, the `store-dir` takes its value. If it does not exist, then it takes `~/.pnpm/store`

Pnpm supports this feature
> Values in the .npmrc files may contain env variables using the ${NAME} syntax. The env variables may also be specified with default values. Using ${NAME-fallback} will return fallback if NAME isn't set. ${NAME:-fallback} will return fallback if NAME isn't set, or is an empty string. —— https://pnpm.io/npmrc

However, After executing "rush install", I found that rush modified the content of the npmrc file. The line "store-dir" was added with the prefix `; MISSING ENVIRONMENT VARIABLE`. This causes this configuration to be invalid.
```ini
; MISSING ENVIRONMENT VARIABLE: store-dir=${PNPM_STORE_DIR:-~/.pnpm/store}
```

## Repro steps

Add `store-dir=${PNPM_STORE_DIR:-~/.pnpm/store}` to the `common/config/rush/.npmrc` file.
Then, execute `rush install` command.

  **Expected result:** 

The content of `common/temp/.npmrc` displays `store-dir=${PNPM_STORE_DIR:-~/.pnpm/store}`

  **Actual result:** 

The content of `common/temp/.npmrc` displays `; MISSING ENVIRONMENT VARIABLE: store-dir=${PNPM_STORE_DIR:-~/.pnpm/store}`

## Details

I located the logical code causing the problem. This code seems to have not considered the fallback syntax.
Actually, I don't understand why the prefix `MISSING ENVIRONMENT VARIABLE` is added here.

https://github.com/microsoft/rushstack/blob/ca5a3a99f92da8519ea64276e3a4c53ca5befb26/libraries/rush-lib/src/utilities/npmrcUtilities.ts#L55-L89

## Standard questions

Please answer these questions to help us investigate your issue more quickly:

| Question | Answer |
| -------- | -------- |
| `@microsoft/rush` globally installed version? | 5.142.0 |
| `rushVersion` from rush.json? | 5.142.0 |
| Operating system? | Mac |
| Would you consider contributing a PR? | Yes |
| Node.js version (`node -v`)? | 20.16.0 |

### PR DESCRIPTION
**Pull request by fzxen**
## Summary

Fixes #5029

## Details

I added a regular expression which copies from @pnpm/config.env-replace/env-replace.ts for identifying fallback syntax.

## How it was tested

I added some test cases in the `libraries/rush-lib/src/utilities/test/npmrcUtilities.test.ts` file to ensure the rush work goes as expected.

### CODE REVIEW

File: libraries/rush-lib/src/utilities/npmrcUtilities.ts [hunk 1]
```diff
@@ -66,11 +46,30 @@ function _trimNpmrcFile(options: {
...
+           *
+           *                                name          fallback
+           * nameString                 ->  nameString    undefined
+           * nameString-fallbackString  ->  nameString    fallbackString
+           * nameString:-fallbackString ->  nameString    fallbackString
+           */
+          const matched: string[] | null = nameWithFallback.match(/([^:-]+)(:?)-(.+)/);
+          // matched: [originStr, variableName, colon, fallback]
+          const name: string = matched?.[1] ?? nameWithFallback;
+          const fallback: string | undefined = matched?.[3];
```
Comment (iclanton) [inline libraries/rush-lib/src/utilities/npmrcUtilities.ts:69]:
```suggestion
          const matched: string[] | null = nameWithFallback.match(/^([^:-]+)(?:\:?-(.+))?$/);
          // matched: [originStr, variableName, colon, fallback]
          const name: string = matched?.[1] ?? nameWithFallback;
          const fallback: string | undefined = matched?.[2];
```

File: libraries/rush-lib/src/utilities/npmrcUtilities.ts [hunk 2]
```diff
@@ -66,11 +46,30 @@ function _trimNpmrcFile(options: {
       const environmentVariables: string[] | null = line.match(expansionRegExp);
```
Comment (iclanton) [inline libraries/rush-lib/src/utilities/npmrcUtilities.ts:46]: Consider replacing `expansionRegExp` with `(?:\$\{)([^\}:-]+)(?:\:?-(.+))?(?:\})`. Then `match[0]` is the env var name and `match[1]` is the fallback value.

File: libraries/rush-lib/src/utilities/test/npmrcUtilities.test.ts [hunk 3]
```diff
@@ -0,0 +1,19 @@
+// Copyright (c) Microsoft Corporation. All rights reserved. Licensed under the MIT license.
+// See LICENSE in the project root for license information.
+
+import { trimNpmrcFileLines } from '../npmrcUtilities';
+
+describe('npmrcUtilities', () => {
+  it(trimNpmrcFileLines.name, () => {
```
Comment (iclanton) [inline libraries/rush-lib/src/utilities/test/npmrcUtilities.test.ts:7]: Add some cases without env vars and with malformed expansion expressions. Also add some multiline cases.

File: libraries/rush-lib/src/utilities/npmrcUtilities.ts (hunk 2 above)
Comment (fzxen) [reply libraries/rush-lib/src/utilities/npmrcUtilities.ts:50]: I found that when encountering multi-environment variables like `var1=${foo}-${bar}`, this regular expression fails.

File: libraries/rush-lib/src/utilities/test/npmrcUtilities.test.ts [hunk 4]
```diff
@@ -0,0 +1,37 @@
...
+    // Multiline
+    expect(trimNpmrcFileLines(['var1=${foo}', 'var2=${bar}'], { foo: 'test' })).toEqual(['var1=${foo}', addMissingEnvPrefix('var2=${bar}')]);
+    expect(trimNpmrcFileLines(['var1=${foo}', 'var2=${bar}'], { foo: 'test', bar: 'test' })).toEqual(['var1=${foo}', 'var2=${bar}']);
+    expect(trimNpmrcFileLines(['var1=${foo}', 'var2=${bar-fallback_value}'], { foo: 'test' })).toEqual(['var1=${foo}', 'var2=${bar-fallback_value}']);
+    expect(trimNpmrcFileLines(['var1=${foo:-fallback_value}', 'var2=${bar-fallback_value}'], {})).toEqual(['var1=${foo:-fallback_value}', 'var2=${bar-fallback_value}']);
+
+    // Malformed
+    expect(trimNpmrcFileLines(['var1=${foo_fallback_value}'], {})).toEqual([addMissingEnvPrefix('var1=${foo_fallback_value}')]);
+    expect(trimNpmrcFileLines(['var1=${foo:fallback_value}'], {})).toEqual([addMissingEnvPrefix('var1=${foo:fallback_value}')]);
+    expect(trimNpmrcFileLines(['var1=${foo:_fallback_value}'], {})).toEqual([addMissingEnvPrefix('var1=${foo:_fallback_value}')]);
```
Comment (iclanton) [inline libraries/rush-lib/src/utilities/test/npmrcUtilities.test.ts:35]:
```suggestion
    // Malformed
    expect(trimNpmrcFileLines(['var1=${foo_fallback_value}'], {})).toEqual([addMissingEnvPrefix('var1=${foo_fallback_value}')]);
    expect(trimNpmrcFileLines(['var1=${foo:fallback_value}'], {})).toEqual([addMissingEnvPrefix('var1=${foo:fallback_value}')]);
    expect(trimNpmrcFileLines(['var1=${foo:_fallback_value}'], {})).toEqual([addMissingEnvPrefix('var1=${foo:_fallback_value}')]);
    expect(trimNpmrcFileLines(['var1=${foo'], {})).toEqual(['var1=${foo']);
    
```

**Review by iclanton — APPROVED** [review]

**Review by D4N14L — APPROVED** [review]

File: common/changes/@microsoft/rush/feature-support_fallback_syntax_in_npmrc_2024-12-04-13-22.json [hunk 5]
```diff
@@ -0,0 +1,10 @@
+{
+  "changes": [
+    {
+      "packageName": "@microsoft/rush",
+      "comment": "Support fallback syntax in .npmrc file",
```
Comment (iclanton) [inline common/changes/@microsoft/rush/feature-support_fallback_syntax_in_npmrc_2024-12-04-13-22.json:5]:
```suggestion
      "comment": "Support fallback syntax in `.npmrc` files if the package manager is PNPM. See https://pnpm.io/npmrc",
```

File: libraries/rush-lib/src/utilities/npmrcUtilities.ts [hunk 6]
```diff
@@ -22,27 +22,55 @@ export interface ILogger {
...
+/**
+ *
+ * @param npmrcFileLines The npmrc file's lines
+ * @param env The environment variables object
+ * @param supportEnvVarFallbackSyntax Whether to support fallback values in the form of `${VAR_NAME:-fallback}`
+ * @returns
+ */
+export function trimNpmrcFileLines(
+  npmrcFileLines: string[],
+  env: NodeJS.ProcessEnv,
```
Comment (D4N14L) [inline libraries/rush-lib/src/utilities/npmrcUtilities.ts:71]: If you're using the `Pick<...>` syntax above, may as well do it here too.

**Review by D4N14L — APPROVED** [review]

File: libraries/rush-lib/src/utilities/npmrcUtilities.ts (hunk 6 above)
Comment (iclanton) [reply libraries/rush-lib/src/utilities/npmrcUtilities.ts:71]: This isn't an object.

**Pull request merged**